File: tp01/01_palindromo_java/Palindromo.java
/**
 * Palindromo - programa que verifica se cadeias de caracteres
 * de entrada sao palindromos e retorna os resultados
 *
 * @author Pedro Sa (742626)
 * @version 1.1
 * @since 2021-08-03
 */
public class Palindromo
{
    /**
     * Este metodo determina se uma cadeia de caracteres configura
     * ou nao um palindromo.
     * @param line Cadeia de caracteres a ser testada.
     * @return true, se for um palindromo; false, caso contrario.
     */
    public static boolean isPalindrome(String line)
    {
        // definir dados
        boolean pal = true;
        int length = line.length();
        int i = 0;
        int j = length - 1;

        while (pal && j > i)
        {
            // testar se caractere na posicao atual e' igual ao
            // caractere na posicao espelhada
            pal = pal && (line.charAt(j) == line.charAt(i));
            i++;
            j--;
        }

        return pal;
    }

    public static void main(String args[])
    {
        // definir dados
        String line;

        // ler a primeira linha
        line = MyIO.readLine();

        while (!(line.charAt(0) == 'F' &&
                 line.charAt(1) == 'I' &&
                 line.charAt(2) == 'M' &&
                 line.length() == 3))
        {
            // chamar metodo para verificacao e saida
            if (isPalindrome(line))
                MyIO.println("SIM");
            else
                MyIO.println("NAO");

            // ler novamente
            line = MyIO.readLine();
        }
    }
}

File: tts/tt02/palindromo/palindromo.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

/**
 * Este metodo determina se uma cadeia de caracteres configura
 * ou nao um palindromo.
 * @param line Cadeia de caracteres a ser testada.
 * @return true, se for um palindromo; false, caso contrario.
 */
bool isPalindrome(char *line)
{
    // definir dados
    bool pal = true;
    int length = strlen(line) - 1;
    int i = 0;
    int j = length - 1;

    while (pal && j > 1)
    {
        // testar se caractere na posicao atual e' igual ao
        // caractere na posicao espelhada
        pal = pal && (line[j] == line[i]);
        i++;
        j--;
    }

    return pal;
}

int main(void)
{
    // definir dados
    char buffer[200];

    // ler entrada
    fgets(buffer, 200, stdin);

    while (!(buffer[0] == 'F' &&
             buffer[1] == 'I' &&
             buffer[2] == 'M' &&
             strlen(buffer) == 4))
    {
        // testar se e' palindromo e mostrar resultado
        if (isPalindrome(buffer))
        {
            printf("SIM\n");
        }
        else
        {
            printf("NAO\n");
        }

        fgets(buffer, 200, stdin);
    }

    return 0;
}

